File: csharp/leet_code/165.cs
/// <summary>
/// This module contains the Solution class which provides a method to compare two version numbers.
/// </summary>
public class Solution {
    /// <summary>
    /// Compares two version numbers.
    /// </summary>
    /// <param name="version1">A string representing the first version number.</param>
    /// <param name="version2">A string representing the second version number.</param>
    /// <returns>
    /// An integer indicating the comparison result.
    /// Returns -1 if the first version is less than the second, 1 if the first version is greater than the second, and 0 if they are equal.
    /// </returns>
    public int CompareVersion(string version1, string version2) {
        // Split the version numbers into arrays
        var split1 = version1.Split(".");
        var split2 = version2.Split(".");
        int i=0;
        int j=0;

        // Iterate over the arrays while there are elements in either of them
        while (i < split1.Length || j < split2.Length) {
            // Get the current element from each array, or 0 if there are no more elements
            int one = i < split1.Length ? Int32.Parse(split1[i]) : 0;
            int two = j < split2.Length ? Int32.Parse(split2[j]) : 0;
            i++; j++;
            // If the elements are equal, continue to the next iteration
            if (one == two) continue;
            // Return the comparison result
            return one < two ? -1 : 1;
        }
        // If all elements have been compared and are equal, return 0
        return 0;
    }
}

File: php/leet_code/165.php
<?php

/**
 * This module contains the Solution class which provides a method to compare two version numbers.
 * Version numbers consist of one or more revisions separated by a dot '.'. This method compares each revision of the version numbers
 * and returns 1 if version1 > version2, -1 if version1 < version2, and 0 if version1 == version2.
 */

class Solution
{
    /**
     * This method compares two version numbers.
     *
     * @param string $version1 A string representing the first version number.
     * @param string $version2 A string representing the second version number.
     * @return int 1 if version1 > version2, -1 if version1 < version2, and 0 if version1 == version2.
     */
    public function compareVersion(string $version1, string $version2): int
    {
        // Initialize the revisions of the version numbers
        $revision1 = '';
        $revision2 = '';

        $i1 = 0;
        $i2 = 0;
        // Iterate over each character in the version numbers
        while (
            $i1 < strlen($version1) ||
            $i2 < strlen($version2)
        ) {

            $char1 = '';
            // If the end of version1 is reached, set the revision to '0' if it's empty
            if ($i1 == strlen($version1)) {
                if ($revision1 == '') {
                    $revision1 = '0';
                }
            } else {
                $char1 = $version1[$i1];
                // If the character is not a dot, append it to the revision
                if ($char1 != '.') {
                    $revision1 .= $char1;
                    $i1++;
                }
            }

            $char2 = '';
            // If the end of version2 is reached, set the revision to '0' if it's empty
            if ($i2 == strlen($version2)) {
                if ($revision2 == '') {
                    $revision2 = '0';
                }
            } else {
                $char2 = $version2[$i2];
                // If the character is not a dot, append it to the revision
                if ($char2 != '.') {
                    $revision2 .= $char2;
                    $i2++;
                }
            }

            // If a dot or the end of the version numbers is reached, compare the revisions
            if (
                ($char1 == '' || $char1 == '.') &&
                ($char2 == '' || $char2 == '.')
            ) {

                // If the revisions are not equal, break the loop
                if ((int)$revision1 != (int)$revision2) {
                    break;
                }

                // Reset the revisions
                $revision1 = '';
                $revision2 = '';

                // Skip the dots
                if ($char1 == '.') {
                    $i1++;
                }

                if ($char2 == '.') {
                    $i2++;
                }

            }

        }

        // Compare the revisions and return the result
        if ((int)$revision1 < (int)$revision2) {
            return -1;
        }

        if ((int)$revision1 > (int)$revision2) {
            return 1;
        }
        return 0;
    }

}
